### PR TITLE
Hierarchical sampling with composite mass model

### PR DESCRIPTION
**Pull request by ajshajib**
This PR adds the feature to consider a composite mass model in kinematic computation, for example, generalized NFW and stars.

### CODE REVIEW

File: hierarc/LensPosterior/base_config.py [hunk 1]
```diff
@@ -48,15 +49,23 @@ def __init__(
         :param anisotropy_model: type of stellar anisotropy model. See details in MamonLokasAnisotropy() class of lenstronomy.GalKin.anisotropy
         :param multi_observations: bool, if True, interprets kwargs_aperture and kwargs_seeing as lists of multiple
          observations
+         :param lens_model_list: keyword argument list of lens model (optional)
```
Comment (ajshajib) [inline hierarc/LensPosterior/base_config.py:52]:
```suggestion
        :param lens_model_list: keyword argument list of lens model (optional)
```

File: hierarc/Sampling/mcmc_sampling.py [hunk 2]
```diff
@@ -8,8 +8,10 @@ class MCMCSampler(object):
 
     def __init__(self, *args, **kwargs):
         """Initialise the classes of the chain and for parameter options :param args:
-        positional arguments for the CosmoLikelihood() instance :param kwargs: keyword
-        arguments for the CosmoLikelihood() instance."""
+
+        positional arguments for the CosmoLikelihood() instance
+        :param kwargs: keyword arguments for the CosmoLikelihood() instance.
```
Comment (ajshajib) [inline hierarc/Sampling/mcmc_sampling.py:13]:
```suggestion

        :param args: positional arguments for the CosmoLikelihood() instance
        :param kwargs: keyword arguments for the CosmoLikelihood() instance.
```

File: hierarc/Sampling/mcmc_sampling.py [hunk 3]
```diff
@@ -8,8 +8,10 @@ class MCMCSampler(object):
 
     def __init__(self, *args, **kwargs):
         """Initialise the classes of the chain and for parameter options :param args:
```
Comment (ajshajib) [inline hierarc/Sampling/mcmc_sampling.py:10]:
```suggestion
        """Initialise the classes of the chain and for parameter options
```

File: hierarc/Likelihood/KDELikelihood/chain.py [hunk 4]
```diff
@@ -147,6 +147,13 @@ def import_Planck_chain(datapath, kw, probe, params, cosmology, rescale=True):
     :param datapath: (str). Path to the Planck chain :param kw: (str). Planck base
     cosmology keyword. For example, "base" or "base_omegak". See
     https://wiki.cosmos.esa.int/planck-legacy-archive/index.php/Cosmological_Parameters.
+     :param datapath: (str). Path to the Planck chain :param kw: (str). Planck base
+    cosmology keyword. For example, "base" or     "base_omegak". See
+    https://wiki.cosmos.esa.int/planck-legacy-
```
Comment (ajshajib) [inline hierarc/Likelihood/KDELikelihood/chain.py:152]:
```suggestion
    :param datapath: (str). Path to the Planck chain 
    :param kw: (str). Planck base cosmology keyword. For example, "base" or "base_omegak". See
        https://wiki.cosmos.esa.int/planck-legacy-
```

**Comment by ajshajib** [comment]
@sibirrer, looks like [pre-commit.ci](https://pre-commit.ci/) is not inastalled for the repo. In that case, can you kindly do that?

File: hierarc/Likelihood/KDELikelihood/chain.py [hunk 5]
```diff
@@ -147,6 +147,12 @@ def import_Planck_chain(datapath, kw, probe, params, cosmology, rescale=True):
     :param datapath: (str). Path to the Planck chain :param kw: (str). Planck base
     cosmology keyword. For example, "base" or "base_omegak". See
     https://wiki.cosmos.esa.int/planck-legacy-archive/index.php/Cosmological_Parameters.
+    :param datapath: (str). Path to the Planck chain 
+    :param kw: (str). Planck base cosmology keyword. For example, "base" or "base_omegak". See https://wiki.cosmos.esa.int/planck-legacy-
+    :param datapath: (str). Path to the Planck chain
+    :param kw: (str). Planck base cosmology keyword. For example, "base" or
+        "base_omegak". See https://wiki.cosmos.esa.int/planck-legacy-
+        archive/index.php/Cosmological_Parameters.
```
Comment (ajshajib) [inline hierarc/Likelihood/KDELikelihood/chain.py:155]:
```suggestion
        "base_omegak". See https://wiki.cosmos.esa.int/planck-legacy-archive/index.php/Cosmological_Parameters.
```

File: hierarc/LensPosterior/kin_constraints_composite.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+            kwargs_mge_light=None,
+            hernquist_approx=False,
+            sampling_number=sampling_number,
+            num_psf_sampling=num_psf_sampling,
+            num_kin_sampling=num_kin_sampling,
+            multi_observations=multi_observations,
+        )
+
+        if len(rho0_array) != len(r_s_array):
+            raise ValueError("rho0 and r_s must have the same length")
```
Comment (ajshajib) [inline hierarc/LensPosterior/kin_constraints_composite.py:132]:
```suggestion
            raise ValueError("rho0 and r_s must have the same length!")
```

File: test/test_Likelihood/test_LensLikelihood/test_ddt_dd_gauss_likelihood.py [hunk 7]
```diff
@@ -37,7 +37,7 @@ def test_log_likelihood(self):
 
         aniso_scaling = [1 + self.dd_sigma / self.dd_mean]
         lnlog = likelihood.log_likelihood(
-            ddt=self.ddt_mean, dd=self.dd_mean, aniso_scaling=aniso_scaling
+            ddt=self.ddt_mean, dd=self.dd_mean, kin_scaling=aniso_scaling
```
Comment (ajshajib) [inline test/test_Likelihood/test_LensLikelihood/test_ddt_dd_gauss_likelihood.py:40]:
```suggestion
            ddt=self.ddt_mean, dd=self.dd_mean, kin_scaling=kin_scaling
```

File: test/test_Likelihood/test_LensLikelihood/test_ddt_dd_gauss_likelihood.py [hunk 8]
```diff
@@ -37,7 +37,7 @@ def test_log_likelihood(self):
 
         aniso_scaling = [1 + self.dd_sigma / self.dd_mean]
```
Comment (ajshajib) [inline test/test_Likelihood/test_LensLikelihood/test_ddt_dd_gauss_likelihood.py:38]:
```suggestion
        kin_scaling = [1 + self.dd_sigma / self.dd_mean]
```

File: test/test_LensPosterior/test_kin_constraints.py [hunk 9]
```diff
@@ -216,6 +217,7 @@ def test_likelihoodconfiguration_gom(self):
         )
 
         kwargs_likelihood = kin_constraints.hierarchy_configuration(num_sample_model=5)
+        print(kwargs_likelihood)
```
Comment (ajshajib) [inline test/test_LensPosterior/test_kin_constraints.py:220]:
```suggestion
```

File: test/test_LensPosterior/test_kin_constraints.py [hunk 10]
```diff
@@ -109,6 +109,7 @@ def test_likelihoodconfiguration_om(self):
...
```
Comment (ajshajib) [inline test/test_LensPosterior/test_kin_constraints.py:112]:
```suggestion
```

**Review by sibirrer — APPROVED** [review]
Thank you @ajshajib. Unfortunately I don't see a test coverage, but otherwise it good to go

**Comment by ajshajib** [comment]
Thanks @sibirrer for approval, and also to @williyamshoe for his contributions! The test coverage is a little hidden to see, but it's here: https://app.codecov.io/gh/sibirrer/hierArc/pull/20. The above message from `codecov`, which auto-updated after the last commit, says that the new lines are 100% tested and overall coverage increased by 0.12%.

It also gives the notice, which may have to do with the coverage report being a little hard to find.

> :exclamation: Your organization needs to install the [Codecov GitHub app](https://github.com/apps/codecov/installations/select_target) to enable full functionality.

**Pull request merged**